### PR TITLE
Add wrappers for signature precompiles to Sapphire library

### PR DESCRIPTION
**Pull request by jberci**
Wrappers for the precompiles added in oasisprotocol/oasis-sdk#1176.

### CODE REVIEW

**Review by nhynes — APPROVED** [review]

File: contracts/contracts/Sapphire.sol [hunk 1]
```diff
@@ -160,4 +160,144 @@ library Sapphire {
...
         }
     }
+
+    /**
+     * @dev Generate a public/private key pair using the specified method and seed.
+     * @param method The method to use (0 - ed25519, 1 - secp256k1, 2 - sr25519).
+     * @param seed The seed to use for generating the key pair.
+     * @return (publicKey, privateKey) The generated key pair.
+     */
+    function generateKeyPair(uint method, bytes memory seed) internal view returns (bytes memory publicKey, bytes memory privateKey) {
```
Comment (nhynes) [inline contracts/contracts/Sapphire.sol:170]:
```suggestion
    enum KeyKind {
        Unknown,
        Ed25519,
        Secp256k1,
        Sr25519
    }
    
    function generateKeyPair(KeyKind kind, bytes memory seed) internal view returns (bytes memory publicKey, bytes memory privateKey) {
```

File: contracts/contracts/Sapphire.sol [hunk 2]
```diff
@@ -160,4 +160,144 @@ library Sapphire {
...
+
+    /**
+     * @dev Sign a message within the provided context using the specified method, and return the signature.
+     * @param method The method to use (0 - ed25519, 1 - secp256k1, 2 - sr25519).
+     * @param privateKey The private key to use for signing.
+     * @param context The context to add to the message before signing.
+     * @param message The message to sign.
+     * @return signature The resulting signature.
+     */
+    function signMessageWithContext(uint method, bytes memory privateKey, bytes memory context, bytes memory message) internal view returns (bytes memory signature) {
```
Comment (nhynes) [inline contracts/contracts/Sapphire.sol:211]:
```suggestion
    function signMessageWithContext(SigningKey memory key, bytes memory context, bytes memory message) internal view returns (bytes memory signature) {
```
also, this function is probably `pure`? the linter will tell you, I think.

File: contracts/contracts/Sapphire.sol [hunk 3]
```diff
@@ -160,4 +160,144 @@ library Sapphire {
...
+    /**
+     * @dev Verify that the context and message correspond to the given signature.
+     * @param method The method to use (0 - ed25519, 1 - secp256k1, 2 - sr25519).
+     * @param publicKey The public key to use to check the signature.
+     * @param signature The signature to check.
+     * @param context The context to add to the message before checking.
+     * @param message The message to check.
+     * @return result The result of the verification; `True` if the verification succeeded.
+     */
+    function verifySignatureWithContext(uint method, bytes memory publicKey, bytes memory signature, bytes memory context, bytes memory message) internal view returns (bool result) {
```
Comment (nhynes) [inline contracts/contracts/Sapphire.sol:260]:
```suggestion
    function verifySignatureWithContext(SigningAlg alg, bytes memory publicKey, bytes memory signature, bytes memory context, bytes memory message) internal view returns (bool result) {
```

**Comment by nhynes** [comment]
Thanks @jberci! I'll shepherd this PR from here since I made a number of changes to the precompile ABIs.

File: contracts/contracts/Sapphire.sol [hunk 4]
```diff
@@ -6,158 +6,174 @@ pragma solidity ^0.8.0;
...
+    address private constant VERIFY_DIGEST =
+        0x0100000000000000000000000000000000000007;
+
+    enum SigningAlg {
+        // Sentinel for an unset signing alg. Do not pass this into any precompiles.
+        Unknown,
+        // Ed25519 (using SHA-512).
+        Ed25519,
+        // EdDSA using SHA-512/256 and Curve25519 with a domain separator.
+        // Ccan be used to sign transactions for the Oasis consensus layer and SDK paratimes.
```
Comment (kostko) [inline contracts/contracts/Sapphire.sol:30]:
```suggestion
        // Can be used to sign transactions for the Oasis consensus layer and SDK paratimes.
```

File: contracts/contracts/Sapphire.sol [hunk 5]
```diff
@@ -6,158 +6,174 @@ pragma solidity ^0.8.0;
...
+        Ed25519Oasis,
+        // Secp256k1 using SHA-256.
+        Secp256k1,
+        // Secp256k1 using SHA-512/256 with a domain separator.
+        // Ccan be used to sign transactions for the Oasis consensus layer and SDK paratimes.
```
Comment (kostko) [inline contracts/contracts/Sapphire.sol:35]:
```suggestion
        // Can be used to sign transactions for the Oasis consensus layer and SDK paratimes.
```

File: contracts/contracts/Sapphire.sol [hunk 6]
```diff
@@ -6,158 +6,174 @@ pragma solidity ^0.8.0;
...
+        Secp256k1Oasis,
+        // Secp256k1 using Keccak256.
+        // Ccan be used to sign transactions for Ethereum-compatible networks.
```
Comment (kostko) [inline contracts/contracts/Sapphire.sol:38]:
```suggestion
        // Can be used to sign transactions for Ethereum-compatible networks.
```

**Review by kostko — APPROVED** [review]
Rebase.

**Pull request merged**